### PR TITLE
feat: refresh credential

### PR DESCRIPTION
**Pull request by xiaoyijun**
## Summary
This PR supports credential refresh by refresh token.

## TBD
[Implement authentication exceptions according to logto-core](https://linear.app/silverhand/issue/LOG-80/implement-authentication-exceptions-according-to-logto-core)

### CODE REVIEW

File: android/library/src/main/java/io/logto/android/utils/TokenUtil.kt [hunk 1]
```diff
@@ -0,0 +1,11 @@
+package io.logto.android.utils
+
+import io.logto.android.model.Credential
+
+object TokenUtil {
+    private const val MILLISECONDS = 1000L
+
+    fun calcAccessTokenExpiredTime(credential: Credential): Long {
```
Comment (wangsijie) [inline android/library/src/main/java/io/logto/android/utils/TokenUtil.kt:8]: call it `expiresAt`?

File: android/library/src/main/java/io/logto/android/Logto.kt [hunk 2]
```diff
@@ -67,8 +61,82 @@ class Logto(
...
+            if (value != null) {
+                field = credentialCache
+                accessTokenExpiredAt = TokenUtil.calcAccessTokenExpiredTime(value)
+            } else {
+                field = null
+                accessTokenExpiredAt = 0L
+            }
+        }
+
+    private var accessTokenExpiredAt: Long = 0L
```
Comment (wangsijie) [inline android/library/src/main/java/io/logto/android/Logto.kt:126]: `accessTokenExpiresAt` ?

File: android/library/src/main/java/io/logto/android/utils/TokenUtil.kt [hunk 3]
```diff
@@ -0,0 +1,11 @@
+package io.logto.android.utils
+
+import io.logto.android.model.Credential
```
Comment (wangsijie) [inline android/library/src/main/java/io/logto/android/utils/TokenUtil.kt:3]: in js sdk, `/token` response interface is called `TokenParameters`

Comment (wangsijie) [reply android/library/src/main/java/io/logto/android/utils/TokenUtil.kt:3]: and local parsed stuff is called `TokenSet` or `tokens`

**Review by wangsijie — CHANGES_REQUESTED** [review]

**Review by wangsijie — APPROVED** [review]

File: android/library/src/main/java/io/logto/android/Logto.kt [hunk 4]
```diff
@@ -67,8 +61,82 @@ class Logto(
...
+            block(null, credential?.accessToken)
+            return
+        }
+
+        refreshCredential { error, credential ->
+            block(error, credential?.accessToken)
+        }
+    }
+
+    fun refreshCredential(
```
Comment (simeng-li) [inline android/library/src/main/java/io/logto/android/Logto.kt:82]: nit: Per offline discussion, update the `credential` namespace as well as the func name. Thx

File: android/library/src/main/java/io/logto/android/Logto.kt [hunk 5]
```diff
@@ -67,8 +61,82 @@ class Logto(
...
+        }
+
+    private var accessTokenExpiresAt: Long = 0L
+
+    private fun updateCredential(credential: Credential?) {
+        credentialCache = credential
+        if (credential == null) {
+            credentialStorage?.clearCredential()
+        } else {
+            credentialStorage?.saveCredential(credential)
```
Comment (simeng-li) [inline android/library/src/main/java/io/logto/android/Logto.kt:133]: nit: Just wondering is it a valid case for us, that some times we only need to update a single / partial  params ? i.e （only access_token not refresh_token） in your `credential` object? If so feels like we could have a merge logic implemented within this `updateCredential` method.  For those function caller only need to pass in the params that needs to be updated

File: android/library/src/main/java/io/logto/android/Logto.kt (hunk 4 above)
Comment (xiaoyijun) [reply android/library/src/main/java/io/logto/android/Logto.kt:82]: TBD: [186](https://linear.app/silverhand/issue/LOG-186/follow-naming-conventions-with-js-sdk)

**Pull request merged**